File: apps/api/src/onboarding/onboarding.state-machine.ts
import {
  OnboardingChip,
  OnboardingField,
  OnboardingProgress,
  OnboardingStepId,
} from '@nodescope/shared';

export type OnboardingSideEffect =
  | { type: 'SaveNetwork'; payload: SaveNetworkPayload }
  | { type: 'SaveBrowserDevice'; payload: SaveBrowserDevicePayload }
  | { type: 'SaveRouterDevice'; payload: SaveDevicePayload }
  | { type: 'SaveModemDevice'; payload: SaveDevicePayload }
  | { type: 'SaveHomeIp'; payload: { ip: string } }
  | { type: 'GeocodeAddress'; payload: { address: string } };

export interface SaveNetworkPayload {
  name: string;
  homeAddress?: string;
  homeLatitude?: number;
  homeLongitude?: number;
  isp?: string;
  downMbps?: number;
  upMbps?: number;
}

export interface SaveBrowserDevicePayload {
  name: string;
  mobility: 'HOME_ONLY' | 'ROAMS' | 'UNKNOWN';
}

export interface SaveDevicePayload {
  name: string;
  macAddress: string | undefined;
}

export type OnboardingInput =
  | { kind: 'init' }
  | { kind: 'chip'; value: string }
  | { kind: 'fields'; values: Record<string, string | number> };

export interface StepRender {
  chips: OnboardingChip[];
  fields: OnboardingField[];
}

export interface StepResult {
  nextStepId: OnboardingStepId;
  progress: OnboardingProgress;
  sideEffects: OnboardingSideEffect[];
  complete: boolean;
}

const SKIP_CHIP: OnboardingChip = { label: 'Skip', value: 'skip' };

/**
 * Returns the chips + fields the client should display for a given step. The
 * bot message comes from AiService.generateOnboardingMessage — this function
 * is responsible only for the structural UI.
 */
export function renderStep(stepId: OnboardingStepId): StepRender {
  switch (stepId) {
    case 'welcome':
      return { chips: [{ label: "Let's go", value: 'start' }], fields: [] };
    case 'networkName':
      return {
        chips: [],
        fields: [{ key: 'name', kind: 'text', label: 'Network name', placeholder: 'Home', required: true }],
      };
    case 'address':
      return {
        chips: [SKIP_CHIP],
        fields: [{ key: 'address', kind: 'address', label: 'Home address', required: false }],
      };
    case 'browserDeviceName':
      return {
        chips: [],
        fields: [{ key: 'name', kind: 'text', label: 'Name this browser', placeholder: 'My Laptop', required: true }],
      };
    case 'mobility':
      return {
        chips: [
          { label: 'Home only', value: 'HOME_ONLY' },
          { label: 'I take it places', value: 'ROAMS' },
          { label: 'Not sure', value: 'UNKNOWN' },
        ],
        fields: [],
      };
    case 'confirmHomeIp':
      return {
        chips: [
          { label: 'Yes, this is home', value: 'yes' },
          { label: 'No, skip for now', value: 'no' },
        ],
        fields: [],
      };
    case 'routerMac':
      return {
        chips: [SKIP_CHIP],
        fields: [
          { key: 'name', kind: 'text', label: 'Router name', placeholder: 'Main Router', required: true },
          { key: 'macAddress', kind: 'mac', label: 'Router MAC (optional)' },
        ],
      };
    case 'modemMac':
      return {
        chips: [SKIP_CHIP],
        fields: [
          { key: 'name', kind: 'text', label: 'Modem name', placeholder: 'Modem', required: true },
          { key: 'macAddress', kind: 'mac', label: 'Modem MAC (optional)' },
        ],
      };
    case 'isp':
      return {
        chips: [SKIP_CHIP],
        fields: [{ key: 'isp', kind: 'text', label: 'ISP name', placeholder: 'Comcast' }],
      };
    case 'speeds':
      return {
        chips: [SKIP_CHIP],
        fields: [
          { key: 'downMbps', kind: 'number', label: 'Download Mbps' },
          { key: 'upMbps', kind: 'number', label: 'Upload Mbps' },
        ],
      };
    case 'done':
      return { chips: [{ label: 'Close', value: 'close' }], fields: [] };
  }
}

/**
 * Pure state-transition function. Given the current step + cumulative
 * progress + the user's input, returns the next step plus any side effects
 * that the OnboardingService must enact (DB writes, geocode lookups, etc.).
 *
 * No DB / AI / network access in here — everything is data in, data out, so
 * the full tree of paths can be tested without infrastructure.
 */
export function handleStep(
  stepId: OnboardingStepId,
  progress: OnboardingProgress,
  input: OnboardingInput,
): StepResult {
  switch (stepId) {
    case 'welcome':
      return advance(progress, 'networkName', []);

    case 'networkName': {
      const name = readField(input, 'name');
      if (!name) return stay(progress, stepId);
      return advance({ ...progress, networkName: name }, 'address', []);
    }

    case 'address': {
      if (isSkipChip(input)) {
        return advance(progress, 'browserDeviceName', [
          {
            type: 'SaveNetwork',
            payload: { name: progress.networkName! },
          },
        ]);
      }
      const address = readField(input, 'address');
      if (!address) return stay(progress, stepId);
      return advance(
        { ...progress, homeAddress: address },
        'browserDeviceName',
        [
          {
            type: 'SaveNetwork',
            payload: { name: progress.networkName!, homeAddress: address },
          },
          { type: 'GeocodeAddress', payload: { address } },
        ],
      );
    }

    case 'browserDeviceName': {
      const name = readField(input, 'name');
      if (!name) return stay(progress, stepId);
      return advance({ ...progress, browserDeviceName: name }, 'mobility', []);
    }

    case 'mobility': {
      if (input.kind !== 'chip') return stay(progress, stepId);
      const mobility = input.value as 'HOME_ONLY' | 'ROAMS' | 'UNKNOWN';
      if (!['HOME_ONLY', 'ROAMS', 'UNKNOWN'].includes(mobility)) {
        return stay(progress, stepId);
      }
      return advance({ ...progress, mobility }, 'confirmHomeIp', [
        {
          type: 'SaveBrowserDevice',
          payload: { name: progress.browserDeviceName!, mobility },
        },
      ]);
    }

    case 'confirmHomeIp': {
      if (input.kind !== 'chip') return stay(progress, stepId);
      if (input.value === 'yes') {
        return advance(progress, 'routerMac', [
          { type: 'SaveHomeIp', payload: { ip: 'CURRENT_REQUEST' } },
        ]);
      }
      return advance(progress, 'routerMac', []);
    }

    case 'routerMac':
      return handleMacDeviceStep(stepId, progress, input, {
        progressKey: 'routerMac',
        nextStepId: 'modemMac',
        effectType: 'SaveRouterDevice',
      });

    case 'modemMac':
      return handleMacDeviceStep(stepId, progress, input, {
        progressKey: 'modemMac',
        nextStepId: 'isp',
        effectType: 'SaveModemDevice',
      });

    case 'isp': {
      if (isSkipChip(input)) return advance(progress, 'speeds', []);
      const isp = readField(input, 'isp');
      if (!isp) return stay(progress, stepId);
      return advance({ ...progress, isp }, 'speeds', [
        { type: 'SaveNetwork', payload: { name: progress.networkName!, isp } },
      ]);
    }

    case 'speeds': {
      if (isSkipChip(input)) {
        return advance(progress, 'done', [], true);
      }
      const downMbps = readNumberField(input, 'downMbps');
      const upMbps = readNumberField(input, 'upMbps');
      if (downMbps === undefined && upMbps === undefined) {
        return stay(progress, stepId);
      }
      return advance(
        { ...progress, downMbps, upMbps },
        'done',
        [
          {
            type: 'SaveNetwork',
            payload: { name: progress.networkName!, downMbps, upMbps },
          },
        ],
        true,
      );
    }

    case 'done':
      return { nextStepId: 'done', progress, sideEffects: [], complete: true };
  }
}

/**
 * Shared transition logic for the routerMac and modemMac steps, which are
 * identical apart from the progress key they record, the step they advance to,
 * and the save side effect they emit.
 */
function handleMacDeviceStep(
  stepId: OnboardingStepId,
  progress: OnboardingProgress,
  input: OnboardingInput,
  opts: {
    progressKey: 'routerMac' | 'modemMac';
    nextStepId: OnboardingStepId;
    effectType: 'SaveRouterDevice' | 'SaveModemDevice';
  },
): StepResult {
  if (isSkipChip(input)) return advance(progress, opts.nextStepId, []);
  const name = readField(input, 'name');
  const macAddress = readField(input, 'macAddress');
  if (!name) return stay(progress, stepId);
  return advance({ ...progress, [opts.progressKey]: macAddress }, opts.nextStepId, [
    // SaveRouterDevice and SaveModemDevice carry the identical SaveDevicePayload,
    // so the union-typed `type` needs a cast to land in the discriminated union.
    { type: opts.effectType, payload: { name, macAddress } } as OnboardingSideEffect,
  ]);
}

function advance(
  progress: OnboardingProgress,
  nextStepId: OnboardingStepId,
  sideEffects: OnboardingSideEffect[],
  complete = false,
): StepResult {
  return { nextStepId, progress, sideEffects, complete };
}

function stay(progress: OnboardingProgress, stepId: OnboardingStepId): StepResult {
  return { nextStepId: stepId, progress, sideEffects: [], complete: false };
}

function readField(input: OnboardingInput, key: string): string | undefined {
  if (input.kind !== 'fields') return undefined;
  const raw = input.values[key];
  if (typeof raw !== 'string') return undefined;
  const trimmed = raw.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function readNumberField(input: OnboardingInput, key: string): number | undefined {
  if (input.kind !== 'fields') return undefined;
  const raw = input.values[key];
  if (typeof raw === 'number' && Number.isFinite(raw)) return raw;
  if (typeof raw === 'string' && raw.trim().length > 0) {
    const n = Number(raw);
    return Number.isFinite(n) ? n : undefined;
  }
  return undefined;
}

function isSkipChip(input: OnboardingInput): boolean {
  return input.kind === 'chip' && input.value === 'skip';
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Domain/Onboarding/OnboardingStateMachine.cs
using System.Globalization;

namespace NodeScope.Modules.Inventory.Domain.Onboarding;

/// <summary>The wizard's steps, in the order the happy path walks them.</summary>
public enum OnboardingStepId
{
    Welcome,
    NetworkName,
    Address,
    BrowserDeviceName,
    Mobility,
    ConfirmHomeIp,
    RouterMac,
    ModemMac,
    Isp,
    Speeds,
    Done,
}

/// <summary>Wire labels for <see cref="OnboardingStepId"/> (camelCase, as the client reads them).</summary>
public static class OnboardingStepIds
{
    private static readonly (OnboardingStepId Value, string Label)[] Table =
    [
        (OnboardingStepId.Welcome, "welcome"),
        (OnboardingStepId.NetworkName, "networkName"),
        (OnboardingStepId.Address, "address"),
        (OnboardingStepId.BrowserDeviceName, "browserDeviceName"),
        (OnboardingStepId.Mobility, "mobility"),
        (OnboardingStepId.ConfirmHomeIp, "confirmHomeIp"),
        (OnboardingStepId.RouterMac, "routerMac"),
        (OnboardingStepId.ModemMac, "modemMac"),
        (OnboardingStepId.Isp, "isp"),
        (OnboardingStepId.Speeds, "speeds"),
        (OnboardingStepId.Done, "done"),
    ];

    public static string Of(OnboardingStepId value) =>
        Table.First(entry => entry.Value == value).Label;

    public static OnboardingStepId? TryParse(string? label) =>
        Table.Where(entry => entry.Label == label).Select(entry => (OnboardingStepId?)entry.Value).FirstOrDefault();
}

/// <summary>What the wizard has collected so far. Absent members are omitted from the wire.</summary>
public sealed record OnboardingProgress
{
    public string? NetworkName { get; init; }

    public string? HomeAddress { get; init; }

    public string? BrowserDeviceName { get; init; }

    public string? Mobility { get; init; }

    public string? RouterMac { get; init; }

    public string? ModemMac { get; init; }

    public string? Isp { get; init; }

    public double? DownMbps { get; init; }

    public double? UpMbps { get; init; }
}

/// <summary>A selectable chip in the step's UI.</summary>
public sealed record OnboardingChip(string Label, string Value);

/// <summary>An input field in the step's UI.</summary>
public sealed record OnboardingField(
    string Key,
    string Kind,
    string Label,
    string? Placeholder = null,
    bool? Required = null);

/// <summary>The structural UI of a step; the bot message comes from the narrator.</summary>
public sealed record StepRender(IReadOnlyList<OnboardingChip> Chips, IReadOnlyList<OnboardingField> Fields);

/// <summary>The user's answer to a step.</summary>
public abstract record OnboardingInput;

/// <summary>No answer - the client is asking for the current step (the first turn).</summary>
public sealed record InitInput : OnboardingInput;

/// <summary>The user picked a chip.</summary>
public sealed record ChipInput(string Value) : OnboardingInput;

/// <summary>The user submitted the step's fields.</summary>
public sealed record FieldsInput(IReadOnlyDictionary<string, OnboardingFieldValue> Values) : OnboardingInput;

/// <summary>
/// One submitted field value. The wire admits a string or a number, and the distinction is
/// load-bearing: a text field only accepts a string, while a number field also accepts a
/// numeric string.
/// </summary>
public sealed record OnboardingFieldValue(string? Text, double? Number);

/// <summary>A persistence or lookup action the service must enact after a transition.</summary>
public abstract record OnboardingSideEffect;

/// <summary>Create or update the org's network with whatever the wizard has collected.</summary>
public sealed record SaveNetworkEffect(
    string Name,
    string? HomeAddress = null,
    string? Isp = null,
    double? DownMbps = null,
    double? UpMbps = null) : OnboardingSideEffect;

public sealed record SaveBrowserDeviceEffect(string Name, string Mobility) : OnboardingSideEffect;

public sealed record SaveRouterDeviceEffect(string Name, string? MacAddress) : OnboardingSideEffect;

public sealed record SaveModemDeviceEffect(string Name, string? MacAddress) : OnboardingSideEffect;

/// <summary>Stamp the request's IP as the network's home public IP.</summary>
public sealed record SaveHomeIpEffect : OnboardingSideEffect;

public sealed record GeocodeAddressEffect(string Address) : OnboardingSideEffect;

public sealed record StepResult(
    OnboardingStepId NextStepId,
    OnboardingProgress Progress,
    IReadOnlyList<OnboardingSideEffect> SideEffects,
    bool Complete);

/// <summary>
/// The wizard's pure transition and render logic (Node's <c>onboarding.state-machine.ts</c>):
/// data in, data out, so every path is testable without a database, AI, or network.
/// </summary>
public static class OnboardingStateMachine
{
    private static readonly OnboardingChip SkipChip = new("Skip", "skip");

    private static readonly string[] MobilityValues = ["HOME_ONLY", "ROAMS", "UNKNOWN"];

    /// <summary>The chips and fields the client should display for a step.</summary>
    public static StepRender Render(OnboardingStepId stepId) => stepId switch
    {
        OnboardingStepId.Welcome => new StepRender([new OnboardingChip("Let's go", "start")], []),
        OnboardingStepId.NetworkName => new StepRender(
            [],
            [new OnboardingField("name", "text", "Network name", "Home", true)]),
        OnboardingStepId.Address => new StepRender(
            [SkipChip],
            [new OnboardingField("address", "address", "Home address", Required: false)]),
        OnboardingStepId.BrowserDeviceName => new StepRender(
            [],
            [new OnboardingField("name", "text", "Name this browser", "My Laptop", true)]),
        OnboardingStepId.Mobility => new StepRender(
            [
                new OnboardingChip("Home only", "HOME_ONLY"),
                new OnboardingChip("I take it places", "ROAMS"),
                new OnboardingChip("Not sure", "UNKNOWN"),
            ],
            []),
        OnboardingStepId.ConfirmHomeIp => new StepRender(
            [new OnboardingChip("Yes, this is home", "yes"), new OnboardingChip("No, skip for now", "no")],
            []),
        OnboardingStepId.RouterMac => new StepRender(
            [SkipChip],
            [
                new OnboardingField("name", "text", "Router name", "Main Router", true),
                new OnboardingField("macAddress", "mac", "Router MAC (optional)"),
            ]),
        OnboardingStepId.ModemMac => new StepRender(
            [SkipChip],
            [
                new OnboardingField("name", "text", "Modem name", "Modem", true),
                new OnboardingField("macAddress", "mac", "Modem MAC (optional)"),
            ]),
        OnboardingStepId.Isp => new StepRender(
            [SkipChip],
            [new OnboardingField("isp", "text", "ISP name", "Comcast")]),
        OnboardingStepId.Speeds => new StepRender(
            [SkipChip],
            [
                new OnboardingField("downMbps", "number", "Download Mbps"),
                new OnboardingField("upMbps", "number", "Upload Mbps"),
            ]),
        OnboardingStepId.Done => new StepRender([new OnboardingChip("Close", "close")], []),
        _ => throw new ArgumentOutOfRangeException(nameof(stepId)),
    };

    /// <summary>
    /// Applies the user's input to the current step. An input of the wrong kind (or an empty
    /// required field) leaves the wizard where it is rather than failing the request.
    /// </summary>
    public static StepResult Handle(OnboardingStepId stepId, OnboardingProgress progress, OnboardingInput input)
    {
        ArgumentNullException.ThrowIfNull(progress);
        switch (stepId)
        {
            case OnboardingStepId.Welcome:
                return Advance(progress, OnboardingStepId.NetworkName, []);

            case OnboardingStepId.NetworkName:
            {
                var name = ReadField(input, "name");
                return name is null
                    ? Stay(progress, stepId)
                    : Advance(progress with { NetworkName = name }, OnboardingStepId.Address, []);
            }

            case OnboardingStepId.Address:
            {
                if (IsSkip(input))
                {
                    return Advance(
                        progress,
                        OnboardingStepId.BrowserDeviceName,
                        [new SaveNetworkEffect(progress.NetworkName!)]);
                }

                var address = ReadField(input, "address");
                return address is null
                    ? Stay(progress, stepId)
                    : Advance(
                        progress with { HomeAddress = address },
                        OnboardingStepId.BrowserDeviceName,
                        [
                            new SaveNetworkEffect(progress.NetworkName!, HomeAddress: address),
                            new GeocodeAddressEffect(address),
                        ]);
            }

            case OnboardingStepId.BrowserDeviceName:
            {
                var name = ReadField(input, "name");
                return name is null
                    ? Stay(progress, stepId)
                    : Advance(progress with { BrowserDeviceName = name }, OnboardingStepId.Mobility, []);
            }

            case OnboardingStepId.Mobility:
            {
                if (input is not ChipInput chip || !MobilityValues.Contains(chip.Value, StringComparer.Ordinal))
                {
                    return Stay(progress, stepId);
                }

                return Advance(
                    progress with { Mobility = chip.Value },
                    OnboardingStepId.ConfirmHomeIp,
                    [new SaveBrowserDeviceEffect(progress.BrowserDeviceName!, chip.Value)]);
            }

            case OnboardingStepId.ConfirmHomeIp:
            {
                if (input is not ChipInput confirm)
                {
                    return Stay(progress, stepId);
                }

                return confirm.Value == "yes"
                    ? Advance(progress, OnboardingStepId.RouterMac, [new SaveHomeIpEffect()])
                    : Advance(progress, OnboardingStepId.RouterMac, []);
            }

            case OnboardingStepId.RouterMac:
                return HandleMacDevice(
                    stepId, progress, input, OnboardingStepId.ModemMac,
                    (mac, next) => next with { RouterMac = mac },
                    (name, mac) => new SaveRouterDeviceEffect(name, mac));

            case OnboardingStepId.ModemMac:
                return HandleMacDevice(
                    stepId, progress, input, OnboardingStepId.Isp,
                    (mac, next) => next with { ModemMac = mac },
                    (name, mac) => new SaveModemDeviceEffect(name, mac));

            case OnboardingStepId.Isp:
            {
                if (IsSkip(input))
                {
                    return Advance(progress, OnboardingStepId.Speeds, []);
                }

                var isp = ReadField(input, "isp");
                return isp is null
                    ? Stay(progress, stepId)
                    : Advance(
                        progress with { Isp = isp },
                        OnboardingStepId.Speeds,
                        [new SaveNetworkEffect(progress.NetworkName!, Isp: isp)]);
            }

            case OnboardingStepId.Speeds:
            {
                if (IsSkip(input))
                {
                    return Advance(progress, OnboardingStepId.Done, [], complete: true);
                }

                var downMbps = ReadNumberField(input, "downMbps");
                var upMbps = ReadNumberField(input, "upMbps");
                if (downMbps is null && upMbps is null)
                {
                    return Stay(progress, stepId);
                }

                return Advance(
                    progress with { DownMbps = downMbps, UpMbps = upMbps },
                    OnboardingStepId.Done,
                    [new SaveNetworkEffect(progress.NetworkName!, DownMbps: downMbps, UpMbps: upMbps)],
                    complete: true);
            }

            case OnboardingStepId.Done:
                return new StepResult(OnboardingStepId.Done, progress, [], true);

            default:
                throw new ArgumentOutOfRangeException(nameof(stepId));
        }
    }

    /// <summary>
    /// The router and modem steps differ only in which progress key they record, where they go
    /// next, and which save effect they emit.
    /// </summary>
    private static StepResult HandleMacDevice(
        OnboardingStepId stepId,
        OnboardingProgress progress,
        OnboardingInput input,
        OnboardingStepId nextStepId,
        Func<string?, OnboardingProgress, OnboardingProgress> recordMac,
        Func<string, string?, OnboardingSideEffect> saveEffect)
    {
        if (IsSkip(input))
        {
            return Advance(progress, nextStepId, []);
        }

        var name = ReadField(input, "name");
        if (name is null)
        {
            return Stay(progress, stepId);
        }

        var macAddress = ReadField(input, "macAddress");
        return Advance(recordMac(macAddress, progress), nextStepId, [saveEffect(name, macAddress)]);
    }

    private static StepResult Advance(
        OnboardingProgress progress,
        OnboardingStepId nextStepId,
        IReadOnlyList<OnboardingSideEffect> sideEffects,
        bool complete = false) =>
        new(nextStepId, progress, sideEffects, complete);

    private static StepResult Stay(OnboardingProgress progress, OnboardingStepId stepId) =>
        new(stepId, progress, [], false);

    private static string? ReadField(OnboardingInput input, string key)
    {
        if (input is not FieldsInput fields
            || !fields.Values.TryGetValue(key, out var raw)
            || raw.Text is null)
        {
            return null;
        }

        var trimmed = raw.Text.Trim();
        return trimmed.Length > 0 ? trimmed : null;
    }

    private static double? ReadNumberField(OnboardingInput input, string key)
    {
        if (input is not FieldsInput fields || !fields.Values.TryGetValue(key, out var raw))
        {
            return null;
        }

        if (raw.Number is { } number)
        {
            return double.IsFinite(number) ? number : null;
        }

        return raw.Text is { } text
            && double.TryParse(text.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed)
            && double.IsFinite(parsed)
                ? parsed
                : null;
    }

    private static bool IsSkip(OnboardingInput input) => input is ChipInput { Value: "skip" };
}
